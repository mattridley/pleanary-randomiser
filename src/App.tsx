import React from 'react';

import Button from './components/Button';
import Cell from './components/Cell';
import Root from './components/Root';
import Table from './components/Table';

import useRandomiser from './hooks/useRandomiser';

import cells from './data/cells.json';

function App() {
  const { selectedCell, isSpinning, spin } = useRandomiser(cells);

  return (
    <Root>
      <Table>
        <tbody>
          {cells.map((row, rowIdx) => (
            <tr key={rowIdx}>
              {row.map((text: string, colIdx, cols) => {
                const cellIdx = rowIdx * cols.length + colIdx;
                return (
                  <Cell key={cellIdx} isSelected={selectedCell === cellIdx} cellIdx={cellIdx}>
                    {text}
                  </Cell>
                );
              })}
            </tr>
          ))}
        </tbody>
      </Table>
      <Button onClick={spin} disabled={isSpinning}>
        SPIN
      </Button>
    </Root>
  );
}

export default App;
