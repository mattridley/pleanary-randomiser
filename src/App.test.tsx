import React from 'react';
import { act } from 'react-dom/test-utils';
import { render, fireEvent, cleanup, RenderResult } from 'react-testing-library';
import 'jest-dom/extend-expect';

import App from './App';

describe('<App />', () => {
  let app: RenderResult;

  beforeEach(() => {
    jest.useFakeTimers();
    app = render(<App />);
  });

  afterEach(() => {
    cleanup();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it('should render correctly', () => expect(app.asFragment()).toMatchSnapshot());

  it('should disable the spin button when spinning', () => {
    const button = app.getByText('SPIN');

    expect(button).not.toBeDisabled();
    act(() => {
      fireEvent.click(button)
    });
    expect(button).toBeDisabled();
  });

  it('should select a new cell every 350 ms', () => {
    act(() => {
      fireEvent.click(app.getByText('SPIN'))
    });

    expect(() => app.getByTestId('selected')).toThrowError('Unable to find an element by: [data-testid="selected"]');
    jest.advanceTimersByTime(350);
    expect(() => app.getByTestId('selected')).not.toThrow();
  });

  it('should stop spinning after 2700ms', () => {
    const button = app.getByText('SPIN');

    expect(button).not.toBeDisabled();
    act(() => {
      fireEvent.click(button)
    });

    jest.advanceTimersByTime(2699);
    expect(button).toBeDisabled();
    jest.advanceTimersByTime(1);
    expect(button).not.toBeDisabled();
  });
});
