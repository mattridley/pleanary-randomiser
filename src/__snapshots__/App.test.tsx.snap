// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-17t95uv-root"
  >
    <table
      class="css-1pnottq-Table et0qlwd0"
    >
      <tbody>
        <tr>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              Explain what you learnt today
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What would you do differently next time?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              List 3 key words from today
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What skills have you developed this lesson?
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              How do you know you have progressed?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What has really stuck with you today?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              Describe the lesson using three adjectives
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What do you need to practice more?
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              Is there anything you did not understand today?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              Have you learnt something new this lesson?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What was your contribution to the lesson?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              How could today's lesson be improved?
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              Is there a key fact you remember from today?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What question(s) do you still have?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              How would you describe your effort this lesson?
            </div>
          </td>
          <td
            class="css-b1baov-Cell"
            data-testid="false"
          >
            <div
              class="css-1ez1ods-content"
            >
              What went well today for you?
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="css-ae2djk-Button e1gz2b3x0"
    >
      SPIN
    </button>
  </div>
</DocumentFragment>
`;
